presto-orc/src/main/java/com/facebook/presto/orc/StripeReader.java

```diff
@@ -330,12 +336,20 @@ private boolean addIncludedStreams(Map<Integer, ColumnEncoding> columnEncodings,
             SharedBuffer sharedDecompressionBuffer)
             throws IOException
     {
-        //
-        // Note: this code does not use the Java 8 stream APIs to avoid any extra object allocation
-        //
+        // get index streams from the stripe meta cache
+        Map<StreamId, OrcDataSourceInput> cachedInputs = getCachedIndexStreams(stripeId, diskRanges);
 
-        // read ranges
-        Map<StreamId, OrcDataSourceInput> streamsData = stripeMetadataSource.getInputs(orcDataSource, stripeId, diskRanges, cacheable);
+        // get the rest of the streams
+        Map<StreamId, OrcDataSourceInput> readInputs = stripeMetadataSource.getInputs(orcDataSource,
```

Comment:
is this the only place that indices streams are fetched?

Reply:
I think so, I didn't find any other obvious place that loads index streams.

Reply:
DWRF doesn't have stripe level stats. So one way of doing stats based filtering was:
1. read footer (one IO), get project indices
2. read indices (another IO)
With indices preloaded with footer, stats based filtering can use that directly. But seems like index reading and data reading is mixed up together with disk range planning, etc. So we may be able to do some optimization there.

Reply:
Presto uses lazy loading for streams, should be fine.

Reply:
> Though Presto uses LazyStream, last Index Stream read could be merged with first data block read and we might still read some data blocks. Though these are corner cases, we can improve these use case later.

I'm sorry I don't quite get what you meant by this. If the file have index cache then we won't read index streams from the source, we will pick them up from the index cache.